File: src/formatters.ts
import type { ArxivPaper, CnkiPaper, DblpBibtexResult, DblpPublication, Paper } from "./types.js";

function buildYearFilterMessage(min?: number, max?: number): string {
  if (min === undefined && max === undefined) {
    return "";
  }

  return ` in year range (${min ?? "earliest"}-${max ?? "latest"})`;
}

function formatPaperDateYear(paper: Paper): number | undefined {
  if (!paper.publishedDate) {
    return undefined;
  }

  const year = new Date(paper.publishedDate).getUTCFullYear();
  return year > 1900 ? year : undefined;
}

export function formatIacrSearchResponse(
  papers: Paper[],
  query: string,
  yearMin?: number,
  yearMax?: number,
): string {
  if (papers.length === 0) {
    return `No papers found for query: ${query}${buildYearFilterMessage(yearMin, yearMax)}`;
  }

  const lines = [
    `Found ${papers.length} IACR papers for query '${query}'${buildYearFilterMessage(yearMin, yearMax)}:`,
    "",
  ];

  for (const [index, paper] of papers.entries()) {
    lines.push(`${index + 1}. **${paper.title}**`);
    lines.push(`   - Paper ID: ${paper.paperId}`);
    lines.push(`   - Authors: ${paper.authors.join(", ")}`);
    lines.push(`   - URL: ${paper.url}`);
    lines.push(`   - PDF: ${paper.pdfUrl}`);

    if (paper.categories.length > 0) {
      lines.push(`   - Categories: ${paper.categories.join(", ")}`);
    }

    if (paper.keywords.length > 0) {
      lines.push(`   - Keywords: ${paper.keywords.join(", ")}`);
    }

    if (paper.abstract) {
      lines.push(`   - Abstract: ${paper.abstract}`);
    }

    lines.push("");
  }

  return lines.join("\n");
}

type DblpFilterOptions = {
  yearFrom?: number;
  yearTo?: number;
  venueFilter?: string;
  includeBibtex?: boolean;
};

function buildDblpFilterMessage(options: DblpFilterOptions): string {
  const filters: string[] = [];

  if (options.yearFrom !== undefined || options.yearTo !== undefined) {
    filters.push(`year range (${options.yearFrom ?? "earliest"}-${options.yearTo ?? "latest"})`);
  }

  if (options.venueFilter) {
    filters.push(`venue '${options.venueFilter}'`);
  }

  return filters.length > 0 ? ` with filters: ${filters.join(", ")}` : "";
}

export function formatDblpSearchResponse(
  results: Array<DblpPublication | DblpBibtexResult>,
  query: string,
  options: DblpFilterOptions,
): string {
  if (results.length === 0) {
    return `No papers found for query: ${query}${buildDblpFilterMessage(options)}`;
  }

  if (options.includeBibtex) {
    const lines = [`Found ${results.length} DBLP BibTeX entries for query '${query}'${buildDblpFilterMessage(options)}:`, ""];

    for (const [index, result] of results.entries()) {
      const bibtex = result as DblpBibtexResult;
      lines.push(`${index + 1}. DBLP Key: ${bibtex.dblpKey}`);
      lines.push(`\`\`\`bibtex`);
      lines.push(bibtex.bibtex);
      lines.push(`\`\`\``);
      lines.push("");
    }

    return lines.join("\n");
  }

  const lines = [`Found ${results.length} DBLP papers for query '${query}'${buildDblpFilterMessage(options)}:`, ""];

  for (const [index, result] of results.entries()) {
    const publication = result as DblpPublication;
    lines.push(`${index + 1}. **${publication.title || "Untitled"}**`);
    lines.push(`   - DBLP Key: ${publication.dblpKey}`);
    lines.push(`   - Authors: ${publication.authors.join(", ")}`);

    if (publication.venue) {
      lines.push(`   - Venue: ${publication.venue}`);
    }

    if (publication.year) {
      lines.push(`   - Year: ${publication.year}`);
    }

    if (publication.doi) {
      lines.push(`   - DOI: ${publication.doi}`);
    }

    if (publication.url) {
      lines.push(`   - URL: ${publication.url}`);
    }

    lines.push("");
  }

  return lines.join("\n");
}

export function formatGoogleScholarSearchResponse(
  papers: Paper[],
  query: string,
  yearLow?: number,
  yearHigh?: number,
): string {
  if (papers.length === 0) {
    return `No papers found for query: ${query}${buildYearFilterMessage(yearLow, yearHigh)}`;
  }

  const lines = [
    `Found ${papers.length} Google Scholar papers for query '${query}'${buildYearFilterMessage(yearLow, yearHigh)}:`,
    "",
  ];

  for (const [index, paper] of papers.entries()) {
    lines.push(`${index + 1}. **${paper.title}**`);
    lines.push(`   - Authors: ${paper.authors.join(", ")}`);

    if (paper.citations > 0) {
      lines.push(`   - Citations: ${paper.citations}`);
    }

    const year = formatPaperDateYear(paper);
    if (year) {
      lines.push(`   - Year: ${year}`);
    }

    if (paper.url) {
      lines.push(`   - URL: ${paper.url}`);
    }

    if (paper.abstract) {
      lines.push(`   - Abstract: ${paper.abstract}`);
    }

    lines.push("");
  }

  return lines.join("\n");
}

type ArxivFilterOptions = {
  dateFrom?: string;
  dateTo?: string;
  categories?: string[];
  sortBy?: "relevance" | "date";
};

function buildArxivFilterMessage(options: ArxivFilterOptions): string {
  const filters: string[] = [];
  if (options.dateFrom || options.dateTo) {
    filters.push(`date range (${options.dateFrom ?? "earliest"} to ${options.dateTo ?? "latest"})`);
  }
  if (options.categories && options.categories.length > 0) {
    filters.push(`categories [${options.categories.join(", ")}]`);
  }
  return filters.length > 0 ? ` with filters: ${filters.join(", ")}` : "";
}

export function formatArxivSearchResponse(
  papers: ArxivPaper[],
  query: string,
  options: ArxivFilterOptions = {},
): string {
  if (papers.length === 0) {
    return `No arXiv papers found for query: ${query}${buildArxivFilterMessage(options)}`;
  }

  const lines = [
    `Found ${papers.length} arXiv papers for query '${query}'${buildArxivFilterMessage(options)}:`,
    "",
  ];

  for (const [index, paper] of papers.entries()) {
    lines.push(`${index + 1}. **${paper.title || "Untitled"}**`);
    lines.push(`   - Paper ID: ${paper.paperId}`);
    if (paper.authors.length > 0) {
      lines.push(`   - Authors: ${paper.authors.join(", ")}`);
    }
    if (paper.primaryCategory) {
      lines.push(`   - Primary Category: ${paper.primaryCategory}`);
    }
    if (paper.categories.length > 0) {
      lines.push(`   - Categories: ${paper.categories.join(", ")}`);
    }
    if (paper.publishedDate) {
      lines.push(`   - Published: ${paper.publishedDate}`);
    }
    if (paper.doi) {
      lines.push(`   - DOI: ${paper.doi}`);
    }
    if (paper.journalRef) {
      lines.push(`   - Journal Ref: ${paper.journalRef}`);
    }
    if (paper.url) {
      lines.push(`   - URL: ${paper.url}`);
    }
    if (paper.pdfUrl) {
      lines.push(`   - PDF: ${paper.pdfUrl}`);
    }
    if (paper.abstract) {
      lines.push(`   - Abstract: ${paper.abstract}`);
    }
    lines.push("");
  }

  return lines.join("\n");
}

export function formatCnkiSearchResponse(papers: CnkiPaper[], query: string): string {
  if (papers.length === 0) {
    return `No CNKI papers found for query: ${query}`;
  }

  const lines = [`Found ${papers.length} CNKI papers for query '${query}':`, ""];

  for (const [index, paper] of papers.entries()) {
    lines.push(`${index + 1}. **${paper.title || "Untitled"}**`);
    if (paper.authors.length > 0) {
      lines.push(`   - Authors: ${paper.authors.join(", ")}`);
    }
    if (paper.source) {
      lines.push(`   - Source: ${paper.source}`);
    }
    if (paper.href) {
      lines.push(`   - URL: ${paper.href}`);
    }
    lines.push("");
  }

  return lines.join("\n");
}

File: src/apaper_mcp/formatters.py
from datetime import datetime
from typing import Any


def _year_filter(minimum: int | None, maximum: int | None) -> str:
    if minimum is None and maximum is None:
        return ""
    return f" in year range ({minimum if minimum is not None else 'earliest'}-{maximum if maximum is not None else 'latest'})"


def format_iacr_search_response(
    papers: list[dict[str, Any]],
    query: str,
    year_min: int | None = None,
    year_max: int | None = None,
) -> str:
    if not papers:
        return f"No papers found for query: {query}{_year_filter(year_min, year_max)}"
    lines = [
        f"Found {len(papers)} IACR papers for query '{query}'{_year_filter(year_min, year_max)}:",
        "",
    ]
    for index, paper in enumerate(papers, 1):
        lines += [
            f"{index}. **{paper['title']}**",
            f"   - Paper ID: {paper['paper_id']}",
            f"   - Authors: {', '.join(paper['authors'])}",
            f"   - URL: {paper['url']}",
            f"   - PDF: {paper['pdf_url']}",
        ]
        if paper["categories"]:
            lines.append(f"   - Categories: {', '.join(paper['categories'])}")
        if paper["keywords"]:
            lines.append(f"   - Keywords: {', '.join(paper['keywords'])}")
        if paper["abstract"]:
            lines.append(f"   - Abstract: {paper['abstract']}")
        lines.append("")
    return "\n".join(lines)


def format_dblp_search_response(
    results: list[dict[str, Any]], query: str, options: dict[str, Any]
) -> str:
    filters = []
    if options.get("year_from") is not None or options.get("year_to") is not None:
        filters.append(
            f"year range ({options.get('year_from', 'earliest')}-{options.get('year_to', 'latest')})"
        )
    if options.get("venue_filter"):
        filters.append(f"venue '{options['venue_filter']}'")
    suffix = f" with filters: {', '.join(filters)}" if filters else ""
    if not results:
        return f"No papers found for query: {query}{suffix}"
    if options.get("include_bibtex"):
        lines = [
            f"Found {len(results)} DBLP BibTeX entries for query '{query}'{suffix}:",
            "",
        ]
        for index, result in enumerate(results, 1):
            lines += [
                f"{index}. DBLP Key: {result['dblp_key']}",
                "```bibtex",
                result["bibtex"],
                "```",
                "",
            ]
        return "\n".join(lines)
    lines = [f"Found {len(results)} DBLP papers for query '{query}'{suffix}:", ""]
    for index, result in enumerate(results, 1):
        lines += [
            f"{index}. **{result['title'] or 'Untitled'}**",
            f"   - DBLP Key: {result['dblp_key']}",
            f"   - Authors: {', '.join(result['authors'])}",
        ]
        if result.get("venue"):
            lines.append(f"   - Venue: {result['venue']}")
        if result.get("year"):
            lines.append(f"   - Year: {result['year']}")
        if result.get("doi"):
            lines.append(f"   - DOI: {result['doi']}")
        if result.get("url"):
            lines.append(f"   - URL: {result['url']}")
        lines.append("")
    return "\n".join(lines)


def format_google_scholar_search_response(
    papers: list[dict[str, Any]],
    query: str,
    year_low: int | None = None,
    year_high: int | None = None,
) -> str:
    if not papers:
        return f"No papers found for query: {query}{_year_filter(year_low, year_high)}"
    lines = [
        f"Found {len(papers)} Google Scholar papers for query '{query}'{_year_filter(year_low, year_high)}:",
        "",
    ]
    for index, paper in enumerate(papers, 1):
        lines += [
            f"{index}. **{paper['title']}**",
            f"   - Authors: {', '.join(paper['authors'])}",
        ]
        if paper["citations"] > 0:
            lines.append(f"   - Citations: {paper['citations']}")
        if paper.get("published_date"):
            try:
                lines.append(
                    f"   - Year: {datetime.fromisoformat(paper['published_date'].replace('Z', '+00:00')).year}"
                )
            except ValueError:
                pass
        if paper.get("url"):
            lines.append(f"   - URL: {paper['url']}")
        if paper.get("abstract"):
            lines.append(f"   - Abstract: {paper['abstract']}")
        lines.append("")
    return "\n".join(lines)


def format_arxiv_search_response(
    papers: list[dict[str, Any]], query: str, options: dict[str, Any] | None = None
) -> str:
    options = options or {}
    filters = []
    if options.get("date_from") or options.get("date_to"):
        filters.append(
            f"date range ({options.get('date_from', 'earliest')} to {options.get('date_to', 'latest')})"
        )
    if options.get("categories"):
        filters.append(f"categories [{', '.join(options['categories'])}]")
    suffix = f" with filters: {', '.join(filters)}" if filters else ""
    if not papers:
        return f"No arXiv papers found for query: {query}{suffix}"
    lines = [f"Found {len(papers)} arXiv papers for query '{query}'{suffix}:", ""]
    for index, p in enumerate(papers, 1):
        lines += [
            f"{index}. **{p['title'] or 'Untitled'}**",
            f"   - Paper ID: {p['paper_id']}",
        ]
        if p["authors"]:
            lines.append(f"   - Authors: {', '.join(p['authors'])}")
        if p["primary_category"]:
            lines.append(f"   - Primary Category: {p['primary_category']}")
        if p["categories"]:
            lines.append(f"   - Categories: {', '.join(p['categories'])}")
        if p["published_date"]:
            lines.append(f"   - Published: {p['published_date']}")
        if p["doi"]:
            lines.append(f"   - DOI: {p['doi']}")
        if p["journal_ref"]:
            lines.append(f"   - Journal Ref: {p['journal_ref']}")
        if p["url"]:
            lines.append(f"   - URL: {p['url']}")
        if p["pdf_url"]:
            lines.append(f"   - PDF: {p['pdf_url']}")
        if p["abstract"]:
            lines.append(f"   - Abstract: {p['abstract']}")
        lines.append("")
    return "\n".join(lines)


def format_cnki_search_response(papers: list[dict[str, Any]], query: str) -> str:
    if not papers:
        return f"No CNKI papers found for query: {query}"
    lines = [f"Found {len(papers)} CNKI papers for query '{query}':", ""]
    for index, p in enumerate(papers, 1):
        lines.append(f"{index}. **{p['title'] or 'Untitled'}**")
        if p["authors"]:
            lines.append(f"   - Authors: {', '.join(p['authors'])}")
        if p["source"]:
            lines.append(f"   - Source: {p['source']}")
        if p["href"]:
            lines.append(f"   - URL: {p['href']}")
        lines.append("")
    return "\n".join(lines)
